websites/I/Instagram/presence.ts

```diff
@@ -1,37 +1,179 @@
 	)
 		presenceData.details = "Direct Messages";
-	else {
-		// TODO: Check if the page is really a profile
-		const [, pathName] = document.location.pathname.split("/");
-		presenceData.details = "Viewing a profile";
-		presenceData.state = pathName;
+	else if (profileName?.textContent === path[1]) {
+		const profilePicture =
+			document.querySelector<HTMLImageElement>("img._6q-tv");
+
+		presenceData.details = `Viewing a Profile${settings.privacy ? "" : ":"}`;
+		presenceData.state = `${
+			document.querySelector("div.QGPIr h1")?.textContent ?? "Unknown"
+		} (@${profileName.textContent})`;
+
+		if (profilePicture)
+			presenceData.smallImageKey = await getShortURL(profilePicture.src);
+
+		presenceData.buttons = [
+			{
+				label: "View Profile",
+				url: `https://www.instagram.com/${path[1]}`
+			}
+		];
+	}
+
+	if (document.querySelector("div.QY4Ed.P0xOK input.focus-visible")) {
+		presenceData.details = settings.privacy ? "Searching" : "Searching:";
+		presenceData.state = document.querySelector<HTMLInputElement>(
+			"div.QY4Ed.P0xOK input.focus-visible"
+		)?.value;
 	}
+
+	if (settings.privacy) {
+		delete presenceData.state;
+		delete presenceData.endTimestamp;
+		delete presenceData.buttons;
+		delete presenceData.smallImageKey;
+	}
+
 	presence.setActivity(presenceData);
 });
+
+function getDateString(date: Date) {
+	const seconds = Math.abs(Date.now() - date.getTime()) / 1000,
+		minutes = Math.floor(seconds / 60),
+		hours = Math.floor(minutes / 60),
+		days = Math.floor(hours / 24);
+
+	switch (true) {
+		case seconds < 60:
+			return `${Math.floor(seconds)} second${seconds === 1 ? "" : "s"} ago`;
+		case seconds < 3600:
+			return `${minutes} minute${minutes === 1 ? "" : "s"} ago`;
+		case seconds < 86400:
+			return `${hours} hour${hours === 1 ? "" : "s"} ago`;
+		case seconds < 604800:
+			return `${days} day${days === 1 ? "" : "s"} ago`;
+		case seconds < 2419200:
+			return `${days / 7} week${days === 7 ? "" : "s"} ago`;
+		case seconds < 29030400:
+			return `${days / 30} month${days === 30 ? "" : "s"} ago`;
+		default:
+			return `${days / 365} year${days === 365 ? "" : "s"} ago`;
+	}
+}
+
+const shortenedURLs: Record<string, string> = {};
+async function getShortURL(url: string) {
+	if (!url || url.length < 256) return url;
+	if (shortenedURLs[url]) return shortenedURLs[url];
+	try {
+		const pdURL = await (
+			await fetch(`https://pd.premid.app/create/${url}`)
+		).text();
+		shortenedURLs[url] = pdURL;
+		return pdURL;
+	} catch (err) {
+		presence.error(err);
+		return url;
+	}
+}
+
+async function getSettings() {
+	const settings = await Promise.all([
+			await presence.getSetting<boolean>("privacy"),
+			await presence.getSetting<boolean>("elapsedTime"),
+			await presence.getSetting<boolean>("postImage")
+		]),
+		names = ["privacy", "elapsedTime", "postImage"],
+		obj: {
+			[key in typeof names[number]]?: boolean;
+		} = {};
+
+	names.forEach((name, i) => {
+		obj[name] = settings[i];
+	});
+
+	return obj;
+}
```

Comment:
What's the point in this?

Reply:
https://github.com/PreMiD/Presences/pull/5641#discussion_r790271040

Reply:
you dont need this whole function tho, just deconstruct the array instead of putting it in an object